core.js

```diff
@@ -372,114 +412,324 @@ exports.init = function (sbot, config) {
     )
   }
 
-  function encryptContent(keys, content) {
-    if (sbot.box2 && content.recps.every(sbot.box2.supportsBox2)) {
-      const feedState = state[keys.id]
-      return sbot.box2.encryptClassic(
-        keys,
-        content,
-        feedState ? feedState.key : null
-      )
-    } else return ssbKeys.box(content, content.recps)
-  }
-
-  function addImmediately(msgVal, cb) {
+  function addImmediately(nativeMsg, opts, cb) {
     const guard = guardAgainstDuplicateLogs('addImmediately()')
     if (guard) return cb(guard)
+    if (typeof opts === 'function') {
+      cb = opts
+      opts = { encoding: 'js' }
+    } else if (!opts) {
+      opts = { encoding: 'js' }
+    }
+    const feedFormat = findFeedFormatByNameOrNativeMsg(
+      opts.feedFormat,
+      nativeMsg
+    )
+    if (!feedFormat) {
+      // prettier-ignore
+      return cb(new Error('addImmediately() failed because could not find feed format for: ' + nativeMsg))
+    }
 
     onceWhen(
       stateFeedsReady,
       (ready) => ready === true,
       () => {
-        if (Ref.isFeedId(msgVal.author)) {
-          const previous = (state[msgVal.author] || { value: null }).value
-          validate2.validateSingle(hmacKey, msgVal, previous, (err, key) => {
-            // prettier-ignore
-            if (err) return cb(clarify(err, 'classic message validation in addImmediately() failed'))
-            updateState({ key, value: msgVal })
-            log.add(key, msgVal, (err, kvt) => {
-              // prettier-ignore
-              if (err) return cb(clarify(err, 'addImmediately() of a classic message failed in the log'))
-
-              post.set(kvt)
-              cb(null, kvt)
-            })
-          })
-        } else if (SSBURI.isBendyButtV1FeedSSBURI(msgVal.author)) {
-          const previous = (state[msgVal.author] || { value: null }).value
-          const err = bendyButt.validateSingle(msgVal, previous, hmacKey)
+        const feedId = feedFormat.getFeedId(nativeMsg)
+        const prevNativeMsg = state.get(feedId)
+        feedFormat.validateSingle(hmacKey, nativeMsg, prevNativeMsg, (err) => {
           // prettier-ignore
-          if (err) return cb(clarify(err, 'bendy butt message validation in addImmediately() failed'))
-          const key = bendyButt.hash(msgVal)
-          updateState({ key, value: msgVal })
-          log.add(key, msgVal, (err, kvt) => {
+          if (err) return cb(clarify(err, 'addImmediately() failed validation for feed format ' + feedFormat.name))
+          const msgId = feedFormat.getMsgId(nativeMsg)
+          const msgVal = feedFormat.fromNativeMsg(nativeMsg, opts.encoding)
+          state.update(feedId, nativeMsg)
```

Comment:
before this was a kvt, and for classic this is fine, we have the key there. But for nativeMsg we don't. So we need a way to store the key here as well. Will speed up validate.

Reply:
That's a good idea! Some back story to how it ended up like this: buttwoo validation has `previousData`, and that's the array of arrays thing, not the previous KVT. So this led to me having to convert `previousKVT` (in the `state`) to a "nativeMsg" (a `previousData` for buttwoo). I then removed the conversion, so that `state` contains the `nativeMsg` directly, and passes that unconverted to the validation layer.

But indeed, I can add the msgId here.

Reply:
Seems to be mostly fixed by weakmap caching, at least for buttwoo